packages/dds/tree/src/core/edit-manager/editManager.ts

```diff
@@ -113,119 +162,139 @@ export class EditManager<
-		return this.trunk;
+		return getPathFromBase(this.trunk, this.trunkBase);
 	}
 
 	public getLastSequencedChange(): TChangeset {
-		return (this.getLastCommit() ?? fail("No sequenced changes")).changeset;
+		return (this.getLastCommit() ?? fail("No sequenced changes")).change;
 	}
 
 	public getLastCommit(): Commit<TChangeset> | undefined {
-		return this.trunk[this.trunk.length - 1];
+		return this.trunk;
 	}
 
 	public getLocalChanges(): readonly RecursiveReadonly<TChangeset>[] {
-		return this.localChanges.map((change) => change.change);
+		return getPathFromBase(this.localBranch, this.trunk).map((c) => c.change);
 	}
 
-	public addSequencedChange(newCommit: Commit<TChangeset>): Delta.Root {
+	public addSequencedChange(
+		newCommit: Commit<TChangeset>,
+		sequenceNumber: SeqNumber,
+		referenceSequenceNumber: SeqNumber,
+	): Delta.Root {
 		assert(
 			this.localSessionId !== undefined,
 			0x429 /* The session ID should be set before processing changes */,
 		);
 
-		if (this.trunk.length > 0) {
-			const lastSeqNumber = this.trunk[this.trunk.length - 1].seqNumber;
-			assert(
-				newCommit.seqNumber > lastSeqNumber,
-				0x3a2 /* Incoming remote op sequence# <= local collabWindow's currentSequence# */,
-			);
-		}
 		if (newCommit.sessionId === this.localSessionId) {
 			// `newCommit` should correspond to the oldest change in `localChanges`, so we move it into trunk.
 			// `localChanges` are already rebased to the trunk, so we can use the stored change instead of rebasing the
 			// change in the incoming commit.
-			const change =
-				this.localChanges.shift() ??
-				fail(
-					"Received a sequenced change from the local session despite having no local changes",
-				);
-
-			// TODO: The local change may contain references to local revision tags.
-			// When other clients rebase this change, they will instead use the corresponding sequence numbers
-			// instead of this client's local revision numbers. Although this need not inherently cause a divergence
-			// in the state of the document, this is potentially dangerous.
-			// This could be solved by one of the following approaches:
-			// A) Maintain a `Branch` of sequenced changes from this client in addition to `localChanges`,
-			//    and only use the rebased changes from that branch when adding to the trunk.
-			// B) Use a UUID as the revision tag for each changeset instead of a sequence number,
-			//    so that we do not need local revision tags.
-			// C) Strip any references to local changes before adding the change to the trunk.
-			//    This would also have to be done on each branch with references to changes local to that branch.
-			this.trunk.push({
-				...newCommit,
-				changeset: change.change,
-			});
+			const localPath = getPathFromBase(this.localBranch, this.trunk);
+			// Get the first revision in the local branch, and then remove it
+			const { change } = localPath.shift() ?? fail(UNEXPECTED_SEQUENCED_LOCAL_EDIT);
```

Comment:
I think we want the string litteral to be passed directly. This will ensure it gets converted to an error code once Craig turns that on for `fail`.